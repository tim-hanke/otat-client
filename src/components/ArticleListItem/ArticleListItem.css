.ArticleListItem__container {
  position: relative;
  margin-bottom: 40px;
  margin-left: 10px;
  margin-right: 10px;
}

.ArticleListItem {
  background: none;
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: smaller;
  text-decoration: none;
  box-shadow: 0px 5px 12px -1px rgba(0, 0, 0, 0.5);
  transition: box-shadow 0.4s ease-out;
  border-radius: 10px;
}

.ArticleListItem__container:hover .ArticleListItem {
  cursor: pointer;
  box-shadow: 0px 8px 20px -1px rgba(0, 0, 0, 0.5);
}

.ArticleListItem,
.ArticleListItem:visited {
  color: #555;
}

.ArticleListItem__image {
  height: 150px;
  width: 275px;
  border-radius: 10px 10px 0 0;
  background-size: 100%;
  transition: background-size 0.4s ease-out;
  background-position: 50% 50%;
}

/* this does a small zoom in on the article image to
create the illusion of the focused item coming towards the user */
.ArticleListItem__container:hover .ArticleListItem__image {
  background-size: 103%;
}

.ArticleListItem__delete {
  background-color: #fff;
  text-align: center;
  height: 30px;
  width: 34px;
  border: 0px gray;
  border-radius: 9px;
  margin: 6px;
  top: 0;
  right: 0;
  position: absolute;
  font-size: x-large;
  cursor: pointer;
}

.ArticleListItem__details {
  padding: 10px;
  display: flex;
  flex: 1;
  flex-direction: column;
  justify-content: space-between;
  width: 255px;
  text-align: center;
}

.ArticleListItem__details h2 {
  margin: 0;
}

/* a little animation that "jiggles" the element*/
/* it goes back and forth, like in traditional animation */
.jiggly:hover,
.jiggly:focus {
  -webkit-animation: jiggle-center 0.4s ease-in-out both;
  animation: jiggle-center 0.4s ease-out both;
}

@-webkit-keyframes jiggle-center {
  0% {
    -webkit-transform: rotate(0);
    transform: rotate(0);
  }
  20% {
    -webkit-transform: rotate(30deg);
    transform: rotate(30deg);
  }
  40% {
    -webkit-transform: rotate(-20deg);
    transform: rotate(-20deg);
  }
  60% {
    -webkit-transform: rotate(5deg);
    transform: rotate(5deg);
  }
  100% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
}
@keyframes jiggle-center {
  0% {
    -webkit-transform: rotate(0);
    transform: rotate(0);
  }
  20% {
    -webkit-transform: rotate(30deg);
    transform: rotate(30deg);
  }
  40% {
    -webkit-transform: rotate(-20deg);
    transform: rotate(-20deg);
  }
  60% {
    -webkit-transform: rotate(5deg);
    transform: rotate(5deg);
  }
  100% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
}
