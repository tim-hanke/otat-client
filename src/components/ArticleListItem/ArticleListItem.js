import React, { Component } from "react";
import { Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrashAlt } from "@fortawesome/free-regular-svg-icons";
import ArticleApiService from "../../services/article-api-service";
import "./ArticleListItem.css";
import ArticleListContext from "../../contexts/ArticleListContext";

export default class ArticleListItem extends Component {
  static contextType = ArticleListContext;

  state = { error: null };

  handleClickDelete = async () => {
    const { id } = this.props.article;

    try {
      await ArticleApiService.deleteUserArticle(id);
      ArticleApiService.getArticles()
        .then(this.context.setArticleList)
        .catch(this.context.setError);
      this.setState({ error: null });
    } catch (res) {
      this.setState({ error: res.error });
    }
  };

  render() {
    const { article } = this.props;
    const { error } = this.state;

    return (
      <div className="ArticleListItem__container">
        <Link
          to={{
            pathname: article.url,
          }}
          target="_blank"
          className="ArticleListItem"
          rel="noopener noreferrer"
        >
          <div
            className="ArticleListItem__image"
            style={{ backgroundImage: `url(${article.image})` }}
          />

          <div className="ArticleListItem__details">
            <div className="ArticleListItem__text">
              <h2 className="ArticleListItem__title">{article.title}</h2>
              <p className="ArticleListItem__description">
                {truncate(article.description)}
              </p>
            </div>
          </div>
          {error && (
            <div role="alert">
              <p>{error}</p>
            </div>
          )}
        </Link>
        <button
          className="ArticleListItem__delete jiggly"
          onClick={this.handleClickDelete}
          aria-label="delete button"
          title="delete button"
        >
          <FontAwesomeIcon className="logo" icon={faTrashAlt} />
        </button>
      </div>
    );
  }
}

function truncate(text) {
  const words = text.split(" ");

  if (words.length > 40) {
    return words.slice(0, 40).join(" ") + " ...";
  }

  return text;
}
