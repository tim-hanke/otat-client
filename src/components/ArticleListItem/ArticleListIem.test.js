import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import ArticleListItem from "./ArticleListItem";

describe.only("ArticleListItem component", () => {
  const testArticle = {
    id: 1,
    url: "https://www.test1.com",
    title: "First test article!",
    image: "http://placehold.it/500x500",
    description:
      "Lorem ipsum dolor sit amet, consectetur adipisicing elit. Natus consequuntur deserunt commodi, nobis qui inventore corrupti iusto aliquid debitis unde non.Adipisci, pariatur.Molestiae, libero esse hic adipisci autem neque ?",
  };

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <ArticleListItem article={testArticle} />
      </MemoryRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <ArticleListItem article={testArticle} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
