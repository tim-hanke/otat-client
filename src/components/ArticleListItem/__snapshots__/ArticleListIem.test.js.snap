// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleListItem component renders the UI as expected 1`] = `
<div
  className="ArticleListItem__container"
>
  <a
    className="ArticleListItem"
    href="https://www.test1.com"
    onClick={[Function]}
    rel="noopener noreferrer"
    target="_blank"
  >
    <div
      className="ArticleListItem__image"
      style={
        Object {
          "backgroundImage": "url(http://placehold.it/500x500)",
        }
      }
    />
    <div
      className="ArticleListItem__details"
    >
      <div
        className="ArticleListItem__text"
      >
        <h2
          className="ArticleListItem__title"
        >
          First test article!
        </h2>
        <p
          className="ArticleListItem__description"
        >
          Lorem ipsum dolor sit amet, consectetur adipisicing elit. Natus consequuntur deserunt commodi, nobis qui inventore corrupti iusto aliquid debitis unde non.Adipisci, pariatur.Molestiae, libero esse hic adipisci autem neque ?
        </p>
      </div>
    </div>
  </a>
  <button
    aria-label="delete button"
    className="ArticleListItem__delete jiggly"
    onClick={[Function]}
    title="delete button"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-trash-alt fa-w-14 logo"
      data-icon="trash-alt"
      data-prefix="far"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M268 416h24a12 12 0 0 0 12-12V188a12 12 0 0 0-12-12h-24a12 12 0 0 0-12 12v216a12 12 0 0 0 12 12zM432 80h-82.41l-34-56.7A48 48 0 0 0 274.41 0H173.59a48 48 0 0 0-41.16 23.3L98.41 80H16A16 16 0 0 0 0 96v16a16 16 0 0 0 16 16h16v336a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128h16a16 16 0 0 0 16-16V96a16 16 0 0 0-16-16zM171.84 50.91A6 6 0 0 1 177 48h94a6 6 0 0 1 5.15 2.91L293.61 80H154.39zM368 464H80V128h288zm-212-48h24a12 12 0 0 0 12-12V188a12 12 0 0 0-12-12h-24a12 12 0 0 0-12 12v216a12 12 0 0 0 12 12z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
</div>
`;
